### ISSUE TITLE
P0528R3 Atomic Compare-And-Exchange With Padding Bits

### ISSUE THREAD

**Issue by StephanTLavavej** [microsoft/STL#23]
[P0528R3](https://wg21.link/P0528R3) Atomic Compare-And-Exchange With Padding Bits
[P1123R0](https://wg21.link/P1123R0) Atomic Compare-And-Exchange With Padding Bits For `atomic_ref`

This needs compiler support, tracked by Microsoft-internal VSO-778892 and MSVC-PR-233835, which will be available in VS 2019 16.7 Preview 1.

**Comment by AlexGuteniev** [comment]
What is intrinsic name?

**Comment by CaseyCarter** [comment]
Oh, I suppose that would be useful, eh?

The intrinsic is `__builtin_zero_non_value_bits`. It takes a pointer to an object, and zeroes padding bits in the representation of that object when evaluated at runtime. At compiletime / in a constant expression it has no effect. Together, that means we can write something like:
```c++
constexpr atomic(T x) noexcept : stored_value(x) {
  __builtin_zero_non_value_bits(&stored_value);
}
```
without any weird `is_constant_evaluated` dance to avoid the intrinsic in constant-initializers.

(And now I'm wondering what the intrinsic does when pointed at a union whose member types have representations with varying amounts of padding, e.g., `union { char c; int i; double d; }`. It can't know which member is active.)

**Comment by AlexGuteniev** [comment]
> Oh, I suppose that would be useful, eh?

Honestly, this one is not that much useful.

I mean it can be used to implement `atomic<T>::compare_exchange_*`, it can also be used to implement properly `atomic<T>::wait`. 

But it is not useful to implement `atomic_ref<T>::compare_exchange_*` or `atomic_ref<T>::wait`.

Need a better intrinsic.

**Comment by CaseyCarter** [comment]
> But it is not useful to implement `atomic_ref<T>::compare_exchange_*` or `atomic_ref<T>::wait`.

Could we rewrite the representation of the object in memory with padding bits zeroed in `atomic_ref`'s constructor? Without some way to fix the values of the padding bits I don't think these functions can be made lock-free: processors don't have operations that atomically compare value representations.

**Comment by AlexGuteniev** [comment]
> Could we rewrite the representation of the object in memory with padding bits zeroed in `atomic_ref`'s constructor?

It is dangerous. `__builtin_zero_non_value_bits` is not atomic.

I'm also not sure if it is useful. The value can be rewritten later not with this or other `atomic_ref`, but directly by other means.

> Without some way to fix the values of the padding bits I don't think these functions can be made lock-free: processors don't have operations that atomically compare value representations.

At least strong `CAS` can be implemented: on failure, if failure is due to padding bits, consider it spurious, and retry with padding bits set to those from the value observed by CAS.

Probably it is not the best way to implement it generally, so apparently we'll should still go with that intrinsic for `atomic` and special-case `atomic_ref`.

Are test feature macros distinct from `atomic` and `atomic_ref` ?

**Comment by StephanTLavavej** [comment]
https://isocpp.org/std/standing-documents/sd-6-sg10-feature-test-recommendations records all feature-test macros; AFAIK these papers don't have macros.

**Comment by AlexGuteniev** [comment]
Actually, I would prefer detection of padding bits, and if there are, fall back to strong CAS that is retrying on padding bits difference. For both `atomic_ref` and `atomic`.

The point is that with padding bits workaround implemented inside CAS, we don't need to pay for the existence of CAS if we use only constructor, load, store and exchange

Actually, I thing even weak CAS can be made. Just do load before CAS and grab expected padding bits, and move them into the value being compared against.

And as weak CAS is implemented using LL/SC, there is a way to handle padding bits as a part of LL/SC based weak CAS just by altering inner comparison. This applies to strong CAS on LL/SC as well.

**Comment by CaseyCarter** [comment]
> > Could we rewrite the representation of the object in memory with padding bits zeroed in `atomic_ref`'s constructor?
> 
> It is dangerous. `__builtin_zero_non_value_bits` is not atomic.

The last modification to the value denoted by an `atomic_ref` before that `atomic_ref` begins its lifetime is not atomic with or without my suggestion.

> I'm also not sure if it is useful. The value can be rewritten later not with this or other `atomic_ref`, but directly by other means.

Not during the lifetime of an `atomic_ref`, it can't: "While any `atomic_ref` instances exist that reference the `*ptr` object, all accesses to that object shall exclusively occur through those `atomic_ref` instances."

> > > Without some way to fix the values of the padding bits I don't think these functions can be made lock-free: processors don't have operations that atomically compare value representations.
> 
> At least strong `CAS` can be implemented: on failure, if failure is due to padding bits, consider it spurious, and retry with padding bits set to those from the value observed by CAS.

Fair enough; I observe that `__builtin_zero_non_value_bits` is sufficient to implement the "if failure is due to padding bits" part of your suggested CAS behavior (we zero the padding of `expected` on entry, and update it to `actual` if `expected` is equal to a copy of `actual` with padding bits zeroed).

**Comment by AlexGuteniev** [comment]
> Not during the lifetime of an `atomic_ref`, it can't: "While any `atomic_ref` instances exist that reference the `*ptr` object, all accesses to that object shall exclusively occur through those `atomic_ref` instances."

Ok, assume one `atomic_ref` is construction, another is already used. Or two `atomic_ref`s are constructed at the same time.

> Fair enough; I observe that `__builtin_zero_non_value_bits` is sufficient to implement the "if failure is due to padding bits" part of your suggested CAS behavior (we zero the padding of `expected` on entry, and update it to `actual` if `expected` is equal to a copy of `actual` with padding bits zeroed).

I think it is a bit suboptimal for x86 and very suboptimal for ARM.

My suggestion:
1. Implement `uint64_t __builting_get_valu_bit_mask64(void*)` thru `uint8_t __builting_get_valu_bit_mask8(void*)`. For x64, also 128-bit version is needed.
2. For x86/x64, compare this way:  ` (expected ^ observed) & significant_bit_mask`
3. For ARM: Add `_InterlockedCompareExchangeMasked` intrinisc, that would implement masked test in its embedded comparison
4. For future weak ARM #775 : Implement weak version as WeakMasked, not just weak.

**Comment by AlexGuteniev** [comment]
So, would you request another intrinsics, or still want to implement with the one you have?

I suggest new, since with the current:
 * The implementation cannot be optimized for the case when no no-value bits on x86
 * The implementation is very suboptimal on ARM

**Comment by AlexGuteniev** [comment]
More cases against  of `atomic_ref` constructor zeroing bits:
1. consider simultaneous `atomic_ref` to different union members, only active is used, and non-active overwrites padding bits
2. consider a reference to read-only memory mapping, where `atomic_ref` only `load`s data

**Comment by AlexGuteniev** [comment]
The existing intrinsic can make `atomic<T>::compare_exchange_*` safe to padding bits by pessimizing other operations (need to zero padding bits on `store`/`exchange`), and this cost of pessimization is especially high for `atomic_ref<T>::compare_exchange_*` (need to atomically zero padding bits, which takes CAS loop in constructor, since there's no atomic version of `__builtin_zero_non_value_bits`

**Comment by AlexGuteniev** [comment]
My suggestion is to request more suitable intrinsic and to move this feature to `Investigating` state, as a possible implementation with current intrinsic isn't good.

**Comment by BillyONeal** [comment]
>The implementation cannot be optimized for the case when no no-value bits on x86

They can, with `has_unique_value_representations`. That also works for ARM, and makes the padding bits part moot for 99% of things that might be put into an atomic.

>consider a reference to read-only memory mapping, where atomic_ref only loads data

I don't think the 'non-lock-free, non-has_unique_value_representations, read-only memory mapping' edge case is substantial enough that we need to worry much about it.

>For x86/x64, compare this way: (expected ^ observed) & significant_bit_mask

The atomicity of the compare/exchange happens inside the `cmpxchg` instruction which doesn't accept a mask.

**Comment by AlexGuteniev** [comment]
> They can, with `has_unique_value_representations`. That also works for ARM, and makes the padding bits part moot for 99% of things that might be put into an atomic.
> 

Didn't know you have it. Ok, this solves perfromance issue for typical case. 
What is intrinsic name?

> I don't think the 'non-lock-free, non-has_unique_value_representations, read-only memory mapping' edge case is substantial enough that we need to worry much about it.

Ok, but there is also problem of construction of `atomic_ref` where another is already constructed.
This alone sorts out any attempts to zero padding bits on construction and assignment.

Only handling contained entirely in CAS looks like a viable option.

> The atomicity of the compare/exchange happens inside the `cmpxchg` instruction which doesn't accept a mask.

Not sure if you understood what I'm proposing. After failed `ccmpxchg`. analyze failure and if it is due to mismatch in non-value bits, retry with non-value bits copied from the observed value.

Sure this can be done using `__builtin_zero_non_value_bits`. But with direct intrinsic that can obtain mask directly, not zero non-value bits, it can be done more optimal. Agree that if there is has_unique_value_representations`, can 

And atomicity of ARM happens between LL/SC so ARM could have masked CAS.

**Comment by AlexGuteniev** [comment]
Apparently the optimizer is smart enough to obtain non-value mask from `__builtin_zero_non_value_bits`:
```
struct X
{
    long a : 7;
};

template<typename T>
long get_non_value_mask()
{
    long value = 0;
    value = ~value;
    __builtin_zero_non_value_bits(reinterpret_cast<T*>(&value));
    return ~value;
}

int main()
{
    return get_non_value_mask<X>();
}
```

```
00281000  mov         eax,0FFFFFF80h  
00281005  ret 
```

So x86 can be implemented correctly and optimally. For ARM still it could be better if there were intrinsics to accept the mask.

**Comment by AlexGuteniev** [comment]
I will try to prepare a PR this weekend

**Comment by BillyONeal** [comment]
>Didn't know you have it. Ok, this solves perfromance issue for typical case.
> What is intrinsic name?

`std::has_unique_object_representations` https://github.com/microsoft/STL/blob/550713eba27c5803addb6f615d3526400cb2df37/stl/inc/type_traits#L618

Probably best to use it by calling the trait rather than trying to use the intrinsic directly.

**Comment by AlexGuteniev** [comment]
> Probably best to use it by calling the trait rather than trying to use the intrinsic directly.

I'm not sure, see #963 , but for now the resolution is such that yes :)

**Comment by BillyONeal** [comment]
> I'm not sure, see #963 , but for now the resolution is such that yes :)

That's a 'customer experience' question, I mean more in avoidance of compiler bugs

**Comment by AlexGuteniev** [comment]
> Honestly, this one is not that much useful

I take it back.

There is non-lock-free generic atomic, where the current form is useful, and for lock-free it is possible to obtain mask from this, see #1029

**Comment by AdamBucior** [comment]
`std::has_unique_object_representations` is not good for detecting types with no padding bits. For example it equals false for floating-point types. There should be an intrinsic specifically for detecting padding bits.

**Comment by AlexGuteniev** [comment]
@AdamBucior , thanks. I've removed `std::has_unique_object_representations`. 

I've checked that compiler is able to optimize the mask comparison and the loop away if `__builtin_zero_non_value_bits` doesn't do anything.

**Comment by AlexGuteniev** [comment]
Interesting that C++20 fixed padding bits, but not fixed non-unique representation of floating points.

**Comment by BillyONeal** [comment]
> `std::has_unique_object_representations` is not good for detecting types with no padding bits. For example it equals false for floating-point types.

It is sufficient but not necessary: has_unique_object_representations implies no padding bits, but no padding bits does not necessarily imply has_unique_object_representations.

>Interesting that C++20 fixed padding bits, but not fixed non-unique representation of floating points.

Because the intended implementation is to zero the padding before ingesting the non-atomic value into the atomic so that the CPU instruction does the right thing, and there is no such normalization scheme that can be done for NaN-packed floats.

**Comment by AlexGuteniev** [comment]
> Because the intended implementation is to zero the padding before ingesting the non-atomic value

I still don't think so, and not supporting floating is not an argument towards either: everything that is possible by normalization in constructor, possible by normalized comparison, and vice versa.

**Comment by BillyONeal** [comment]
>I still don't think so, and not supporting floating is not an argument towards either: everything that is possible by normalization in constructor, possible by normalized comparison, and vice versa.

You can't normalize the different NaNs, they are different values even though they always compare nonequal; it would be bad for `atomic` to break implementations that rely on NaN packing. But with NaN packing, the user has explicit control over what goes on in there; as opposed to padding in general for which the user has no conforming way to observe or access. The former is a feature of the data type and cannot be normalized, the latter is an artifact of the implementation which can be.

**Comment by AlexGuteniev** [comment]
I was more wondering about 0.0 vs -0.0 situation, where values do compare equal. But I now realized that it was also consistent decision.

I also mean that anything implementable with zeroing in constructor/setters is implementable in masking in comparison (by copying padding bits and retrying, or even by doing masked CAS on ARM), and vice versa, so having NaNs does not count towards "zeroing the padding before ingesting the non-atomic value" implementation.

**Comment by AlexGuteniev** [comment]
```
#if _HAS_CXX20
template <class _Ty>
inline constexpr bool _Cmpxchg_has_padding_bits_v = !has_unique_object_representations_v<_Ty> && !is_floating_point_v<_Ty>;
#endif
```

Can work. It does assume that floating point types don't have pure padding, but this assumption is true for the current implementation, where `sizeof(long double) == 8`

**Comment by AlexGuteniev** [comment]
Note, #1029 will not close this issue fully.
After #2 is addressed, there would be `atomic_ref<16 bytes>` specialization (only for `atomic_ref`, not for `atomic`).

If `atomic_ref` is merged earlier, this may be incorporated into #1029, otherwise, if #1029 goes first, there should be additional later change.

**Comment by StephanTLavavej** [comment]
When features overlap, we generally consider the first one to be done, as long as the second feature is implemented with the overlap. So I think it's fine to close this while #2 is still pending.
